evcache-zipkin-tracing/src/main/java/com/netflix/evcache/EVCacheTracingEventListener.java

```diff
@@ -0,0 +1,161 @@
+        serverGroup = client.getServerGroupName();
+        if (StringUtils.isNotBlank(serverGroup)) {
+          serverGroups.add("\"" + serverGroup + "\"");
+        }
+      }
+      clientSpan.tag(EVCacheTracingTags.SERVER_GROUPS, serverGroups.toString());
+
+      /**
+       * Note - EvCache client creates a hash key if the given canonical key size exceeds 255
+       * characters.
+       *
+       * <p>There have been cases where canonical key size exceeded few megabytes. As caching client
+       * creates a hash of such canonical keys and optimizes the storage in the cache servers, it is
+       * safe to annotate hash key instead of canonical key in such cases.
+       */
+      String hashKey;
+      for (EVCacheKey keyObj : e.getEVCacheKeys()) {
+        hashKey = keyObj.getHashKey();
+        if (StringUtils.isNotBlank(hashKey)) {
+          clientSpan.tag(EVCacheTracingTags.HASH_KEY, hashKey);
+        } else {
+          this.safeTag(clientSpan, EVCacheTracingTags.CANONICAL_KEY, keyObj.getCanonicalKey());
+        }
+      }
+
+      /**
+       * Note - tracer.spanInScope(...) method stores Spans in the thread local object.
+       *
+       * <p>As EVCache write operations are asynchronous and quorum based, we are avoiding attaching
+       * clientSpan with tracer.spanInScope(...) method. Instead, we are storing the clientSpan as
+       * an object in the EVCacheEvent's attributes.
+       */
+      e.setAttribute(CLIENT_SPAN_ATTRIBUTE_KEY, clientSpan);
+    } catch (Exception exception) {
+      logger.error("onStart exception", exception);
+    }
+  }
+
+  @Override
+  public void onComplete(EVCacheEvent e) {
+    try {
+      this.onFinishHelper(e, null);
+    } catch (Exception exception) {
+      logger.error("onComplete exception", exception);
+    }
+  }
+
+  @Override
+  public void onError(EVCacheEvent e, Throwable t) {
+    try {
+      this.onFinishHelper(e, t);
+    } catch (Exception exception) {
+      logger.error("onError exception", exception);
+    }
+  }
+
+  /**
+   * On throttle is not a trace event, but it is used to decide whether to throttle. We don't want
+   * to interfere so always return false.
+   */
+  @Override
+  public boolean onThrottle(EVCacheEvent e) throws EVCacheException {
+    return false;
+  }
+
+  private void onFinishHelper(EVCacheEvent e, Throwable t) {
+    Object clientSpanObj = e.getAttribute(CLIENT_SPAN_ATTRIBUTE_KEY);
+
+    // Return if the previously saved Client Span is null
+    if (clientSpanObj == null) {
+      return;
+    }
+
+    Span clientSpan = (Span) clientSpanObj;
+
+    try {
+      if (t != null) {
+        this.safeTag(clientSpan, EVCacheTracingTags.ERROR, t.toString());
+      }
+
+      String status = e.getStatus();
+      this.safeTag(clientSpan, EVCacheTracingTags.STATUS, status);
+
+      long latency = e.getDurationInMillis();
+      clientSpan.tag(EVCacheTracingTags.LATENCY, String.valueOf(latency));
```

Comment:
Is this latency any different from the implicit span duration? ie. is this duplicating that?

Reply:
EVCache client does async operations for quorum. Latency from EVCacheEvent will be appropriate here as underlying implementation of EVCache threads may change in future.

Reply:
Is this latency tag duplicating the _duration_ attribute that Brave adds to all spans though?